File: typescript/tutorials/tour-of-workflows-typescript/src/workflows/signup-with-retries.ts
import * as restate from "@restatedev/restate-sdk";
import { WorkflowContext } from "@restatedev/restate-sdk";
import { activateUser, createUser, sendWelcomeEmail, User } from "../utils";

export const signupWithRetries = restate.workflow({
  name: "SignupWithRetriesWorkflow",
  handlers: {
    run: async (ctx: WorkflowContext, user: User) => {
      const userId = ctx.key;

      const success = await ctx.run("create", () => createUser(userId, user));
      if (!success) return { success };

      await ctx.run("activate", () => activateUser(userId));

      // <start_retries>
      try {
        const retryPolicy = {
          maxRetryAttempts: 3,
          initialRetryInterval: { seconds: 1 },
        };
        await ctx.run("welcome", () => sendWelcomeEmail(user), retryPolicy);
      } catch (error) {
        // This gets hit on retry exhaustion with a terminal error
        // Log and continue; without letting the workflow fail
        console.error("Failed to send welcome email after retries:", error);
      }
      // <end_retries>
      return { success };
    },
  },
});

File: python/tutorials/tour-of-workflows-python/app/workflows/signup_with_retries.py
import restate
from restate import WorkflowContext, TerminalError
from datetime import timedelta
from app.utils import User, create_user, activate_user, send_welcome_email

signup_with_retries = restate.Workflow("SignupWithRetriesWorkflow")


@signup_with_retries.main()
async def run(ctx: WorkflowContext, user: User) -> bool:
    user_id = ctx.key()

    success = await ctx.run_typed("create", create_user, user_id=user_id, user=user)
    if not success:
        return False

    await ctx.run_typed("activate", activate_user, user_id=user_id)

    # Configure retry policy
    try:
        await ctx.run_typed(
            "welcome",
            send_welcome_email,
            restate.RunOptions(
                max_attempts=3, max_retry_duration=timedelta(seconds=30)
            ),
            user=user,
        )
    except TerminalError as error:
        # This gets hit on retry exhaustion with a terminal error
        # Log and continue; without letting the workflow fail
        print(f"Failed to send welcome email after retries: {error}")

    return True
